### ISSUE TITLE
js interop, do I really need to let the user put the javascript script in the html?

### ISSUE THREAD

**Issue by kasperpeulen** [dart-lang/sdk#25059]
I was trying to write a wrapper for https://github.com/prosemirror/prosemirror for Dart.

Something I've been puzzled with, also with other js interop attempts, do I really need to let the user add the javascript in the html.

Because if you look at the demo of the javascript version, you can see that this is not the way it is done in javascript in this example. He only adds a `demo-built.js` file, where the `demo.js` file looks like this:

```javascript
import {ProseMirror} from "../src/edit/main"
import {Pos, Node, LinkStyle} from "../src/model"
import {fromDOM} from "../src/parse/dom"
import {defaultSchema as schema} from "../src/model"

import "../src/inputrules/autoinput"
import "../src/menu/tooltipmenu"
import "../src/menu/menubar"
import "../src/collab"
```

In the demo-built.js file this then looks like this:
```javascript
var _srcEditMain = require("../src/edit/main");
var _srcModel = require("../src/model");
var _srcParseDom = require("../src/parse/dom");
require("../src/inputrules/autoinput");
require("../src/menu/tooltipmenu");
require("../src/menu/menubar");
require("../src/collab");
```

I want the dart users of my wrapper to do something similar. So no need to import it both in the dart file, and also import it the html. Only importing the dart wrapper, should be enough. Would this be possible? If not, I would like this to add this as a feature request.

**Comment by nex3** [comment]
I'm renaming this, since a number of broader issues have been merged into it and the old name was pretty narrowly-scoped.

**Comment by jmesserly** [comment]
https://github.com/dart-lang/sdk/issues/29808 has good discussion on this.

**Comment by jmesserly** [comment]
Related issue, but this one is about how we export a module for JS consumption, rather than import: https://github.com/dart-lang/sdk/issues/27692

**Comment by jmesserly** [comment]
Here's some JS interop examples: https://github.com/matanlurey/dart_js_interop#dart_js_interop

Could you explain more what you're trying to do with webgl? Are there any particular libraries you're trying to use?

**Comment by CosmicPangolin** [comment]
For sure - I'm using 2dimensions' Flutter runtime for a mobile app, and I want to use their webGL runtime in AngularDart to maximize code+asset reuse. See https://github.com/2d-inc/Nima-WebGL

That lib is littered with import/exports, including imports of external params from gl-matrix.  My understanding was that interop couldn't currently handle this structure, and that I'd need to refactor all that code to make it play nice...buuut I could be missing something, since I'm pretty new at all this :)

**Comment by jmesserly** [comment]
You should be able to load that JS library as normal -- any imports/exports inside of JS code are fine. I believe `webpack` can create a single .js file that links everything together.

Then to use that library from Dart, you'd want to write a .dart file that contains the interfaces:
```dart
@JS('Nima')
library nima_webgl;

// note: Dispatcher class needs to be defined too
// it would be nice to add type annotations, especially for return types

@JS()
class Actor extends Dispatcher {
  external factory Actor();
  external addDependency(a, b);
  external sortDependencies();
  external addDirt(component, value, recurse);
  external update();
  external get root;
  external resolveHierarchy(graphics);
  external dispose(graphics);
  external initialize(graphics);
  external advance(seconds);
  external draw(graphics);
  external getNode(name);
  external getAnimation(name);
  external getAnimationInstance(name);
  external makeInstance();
  external computeAABB();
  external copy(actor);
}
```
Similar thing for any other public classes. That makes Dart aware of the JS classes, so you can use them.

**Comment by Aetet** [comment]
@jmesserly Hey. 
Could you provide repo with predefined webpack+js_interop with ES modules?

**Comment by jmesserly** [comment]
I don't know how to use webpack, sorry! They do have documentation: https://webpack.js.org/concepts/, maybe that will help get you started.

**Comment by MikeMitterer** [comment]
@jmesserly Have you ever tried to use one of the top JS libs in Dart? That's one of Dart's weakest points...

**Comment by jmesserly** [comment]
Yeah, totally. It's less than ideal (to put it mildly). I'm working on making that better.

I would like to spend some time thinking about integration with JS module bundlers and build systems too (to address @Aetet's point above). But first, I want to address the core features of interoperability: being able to nicely express and use JS APIs in Dart and vice versa.

**Comment by nex3** [comment]
While it's possible to work around this by treating `require()` like a global function and assigning it to `self` [as in `node_preamble`](https://github.com/mbullington/node_preamble.dart/blob/7b3425ac52b745a294d1d14c1d64238f12a7fca5/lib/preamble.js#L13), this causes problems with downstream tools (see https://github.com/sass/dart-sass/issues/726). Built-in Dart support would be much better.

**Comment by pouledodue** [comment]
@knownasilya did you have success with a webpacked library?
All I get is TypeError "dart.global.MyClass" is not a constructor
